pkg/util/validation/validation.go

```diff
@@ -242,24 +242,33 @@ var httpHeaderNameRegexp = regexp.MustCompile("^" + httpHeaderNameFmt + "$")
 // definition of a valid header field name (a stricter subset than RFC7230).
 func IsHTTPHeaderName(value string) []string {
 	if !httpHeaderNameRegexp.MatchString(value) {
-		return []string{RegexError(httpHeaderNameFmt, "X-Header-Name")}
+		return []string{RegexError(httpHeaderNameFmt, true, "X-Header-Name")}
 	}
 	return nil
 }
 
-const configMapKeyFmt = "\\.?" + dns1123SubdomainFmt
+const configMapKeyFmt = `\A[.\-_a-zA-Z0-9]+[^.]?\z`
+const doubleDotFmt = `\.{2}?`
+const trailingDotFmt = `\.\z`
 
 var configMapKeyRegexp = regexp.MustCompile("^" + configMapKeyFmt + "$")
+var doubleDotRegexp = regexp.MustCompile(doubleDotFmt)
+var trailingDotRegexp = regexp.MustCompile(trailingDotFmt)
 
-// IsConfigMapKey tests for a string that conforms to the definition of a
-// subdomain in DNS (RFC 1123), except that a leading dot is allowed
+// IsConfigMapKey tests for a string that is a valid key for a ConfigMap or Secret
 func IsConfigMapKey(value string) []string {
 	var errs []string
 	if len(value) > DNS1123SubdomainMaxLength {
 		errs = append(errs, MaxLenError(DNS1123SubdomainMaxLength))
 	}
 	if !configMapKeyRegexp.MatchString(value) {
-		errs = append(errs, RegexError(configMapKeyFmt, "key.name"))
+		errs = append(errs, RegexError(configMapKeyFmt, true, "key.name", "KEY_NAME", "key-name"))
```

Comment:
all of this is harder to read than the existing `validateSubPath` function in `pkg/api/validation/validation.go` which could be moved here (most of it) and called `IsDescendingPath` (or similar) and called from both places.